Pulse reference mechanism

nkanazawa1989 opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This PR adds reference mechanism to `ScheduleBlock`. This feature is necessary for template schedule construction in the Qiskit Experiments calibration module. This mechanism allows following workflow.

```python

with pulse.build(name="main") as main_prog:
     pulse.reference("child", "q0")

with pulse.build() as child_prog:
   pulse.play(my_pulse, pulse.DriveChannel(0))

main_prog.assign_references({("child", "q0"): child_prog})
```
Now pulse programmers can `reference` an external program (subroutine) with arbitrary string tuples, i.e. reference keys. This tuple must be unique within the scope (i.e. same name cannot be used twice in the same builder context for calling different programs). Note that `main_prog` can be constructed without knowing realization of `child`. This allows us to decouple programs defined in the separate library. The `main_prog` can (will) be QPY serialized with/without reference assignment.

### Details and comments

Clue for review

(1) Reference manager
New class `ReferenceManager` is introduced. The manager instance is stored in the schedule block instance. Referenced program is appended to the `self._block` as an instruction, and actual subroutine is added to the reference manager.

(2) Parameter management
In the schedule block `ParameterManager` instance is stored as `self._parameter_manager`. Parameter manager now only manages parameters defined in the current scope (i.e. it is not aware of parameters of called program). Instead, management of parameters is delegated to the manager of the referenced program. By recursively calling the manger of called program, it automatically propagates the parameter assignment through all referenced programs from the root program. New property `self.scoped_parameters` is added to return all parameters with scoped name, i.e. if parameter `A` is defined in the program `B`, parameter name becomes `B.A`. If this is further referenced from program `C` the name becomes `C.B.A`. Note that `Parameter` object is not sensitive to name (it is evaluated with UUID).

TODO

- [x] finalize design
- [x] add more tests
- [x] release note

eggerdj commented:
Here is some high level feedback based on a use case. The setting I am working with is the following:
```python
from qiskit.circuit import Parameter
import qiskit.pulse as pulse

x_amp = Parameter("amp")
sigma = Parameter("s")
x_dur = Parameter("dur")
cr_amp = Parameter("amp")
cr_dur = Parameter("dur")
cr_width = Parameter("w")
c_chan_idx = Parameter("ch0")
u_chan_idx = Parameter("ch0.1")
c_chan = pulse.DriveChannel(c_chan_idx)
u_chan = pulse.ControlChannel(u_chan_idx)

with pulse.build(name="xp") as xp:
    pulse.play(pulse.Gaussian(x_dur, x_amp, sigma), c_chan)
    
with pulse.build(name="cr45p") as cr45p:
    pulse.play(pulse.GaussianSquare(cr_dur, cr_amp, width=cr_width, sigma=sigma), u_chan)
    
with pulse.build(name="cr45m") as cr45m:
    pulse.play(pulse.GaussianSquare(cr_dur, -cr_amp, width=cr_width, sigma=sigma), u_chan)
    
with pulse.build(name="ecr") as ecr:
    with pulse.align_sequential():
        pulse.call(name="cr45p", channels=[u_chan])
        pulse.call(name="xp", channels=[c_chan])
        pulse.call(cr45m)  # make things interesting by mixing ref. styles.
        pulse.call(name="xp", channels=[u_chan])
```
Standard assigning still works fine, e.g.
```python
xp_bind_dict = {x_amp: 0.5, x_dur: 160, sigma: 40, c_chan_idx: 2}

xp.assign_parameters(xp_bind_dict, inplace=False).draw()
```
gives:
![image](https://user-images.githubusercontent.com/38065505/171822455-e3cddb2c-2015-40f7-8e28-473380a1db37.png)

Things now start getting more complex when I try to resolve the references and assign parameters in the ECR schedule. Some extra complexity is expected but currently its seems very hard to resolve and assign. My strategy is to first assign parameters in the references (aside from channel indices) and then to assign the references in the `ecr` schedule.
```python
xp_bind_dict = {x_amp: 0.5, x_dur: 160, sigma: 40}
cr_bind_dict = {cr_amp: 0.123, cr_dur: 540, cr_width: 320, sigma: 40}

ecr2 = ecr.assign_reference(ref_key="xp", schedule=xp.assign_parameters(xp_bind_dict, inplace=False), inplace=False)
ecr3 = ecr2.assign_reference(ref_key="cr45p", schedule=cr45p.assign_parameters(cr_bind_dict, inplace=False), inplace=False)

# After this we still need to assign parameters in cr45m because of our strange choice to mix ref styles.
bind_dict = {
    ecr3.get_parameters("s", scope="ecr.cr45m")[0]: 80,  # Also lose the coupling
    ecr3.get_parameters("dur", scope="ecr.cr45m")[0]: 540,
    ecr3.get_parameters("amp", scope="ecr.cr45m")[0]: -0.2,  # This means that we lose the coupling between cr45p and m
    ecr3.get_parameters("w", scope="ecr.cr45m")[0]: 320,
}

ecr4 = ecr3.assign_parameters(bind_dict, inplace=False)
try:
    ecr4.assign_parameters({c_chan_idx: 2, u_chan_idx: 14}).draw()
except pulse.exceptions.UnassignedDurationError as err:
    print(err.message)
    print(ecr4.parameters)  # Contradicts the error message.
```
produces:
```
All instruction durations should be assigned before creating `Schedule`.Please check `.parameters` to find unassigned parameter objects.
set()
```
I also have the following comments:

* For consistency it would be nice to have `assign_reference` behave like the parameter assignment, i.e., `assign_references({"xp": xp_sched, "cr45p": cr45p}, inplace=False)`
* This framework allows us to easily break links between parameter but I'm not sure there is a way around that.
* It seems that the framework does not capture all unassigned parameters?

nkanazawa1989 commented:
Thanks Daniel for nice test case! I updated and simplified the code to support your test case. As we discussed offline, management of parameter-link between different pulse instances is outside the scope of schedule block. This should be delegated to high-level management tools, such as calibrations. However, the scoping mechanism here should be useful to simplify the logic there; by using `get_parameters` with scope, calibrations doesn't need to be aware of parameters (and uuids), and it can manage parameters with scope + parameter name, given we assign unique name to every schedules.

Here is the functional updates:
- `assign_references` takes a dictionary as in `assign_parameters`
- `.parameters` return all parameters in nested subroutines

And there is an overhaul of `ReferenceManager` in https://github.com/Qiskit/qiskit-terra/pull/8005/commits/a66f3cd8ad652abddc682299f8d4c8e7141adfd2. Now this object is a dict-like with pretty printing;

```python
print(ecr4.references)
```
shows
```
ReferenceManager
  - scope: 'ecr'
  - # of references: 3
  - mappings:
    * 'cr45p': ScheduleBlock(Play(GaussianSquare(duration=540, am..., channels=[uch0.1]
    * 'xp': ScheduleBlock(Play(Gaussian(duration=160, amp=(0.5..., channels=[dch0]
    * 'cr45m': ScheduleBlock(Play(GaussianSquare(duration=540, am..., channels=[uch0.1]
```

which looks much better than before. In addition, you have direct access to subroutine with dict-access

```python
ecr4["cr45p"]
```

eggerdj commented:
Playing around with this a bit more. Feels already better after the update. One thing I noticed is the scoping delimiter is based on the `.` symbol and there is no protection against the case where a parameter name has a `.` in it. This leads to a bit of strange behavior. The case is the following (based on the test case above):
```python
u_chan_idx = Parameter("ch0.1")
u_chan = pulse.ControlChannel(u_chan_idx)

...

with pulse.build(name="cr45p") as cr45p:
    pulse.play(pulse.GaussianSquare(cr_dur, cr_amp, width=cr_width, sigma=sigma), u_chan)

with pulse.build(name="ecr") as ecr:
    with pulse.align_sequential():
        pulse.call(name="cr45p", channels=[u_chan])
        ...
```
I can assign the references doing e.g.
```python
ecr2 = ecr.assign_references({"xp": xp, "cr45p": cr45p}, inplace=False)
```
which seems to work fine. Now, I can print the scoped parameters with `ecr2.scoped_parameters` to get
```
{Parameter(ecr.ch0),
 Parameter(ecr.ch0.1),
 Parameter(ecr.cr45p.amp),
 Parameter(ecr.cr45p.dur),
 Parameter(ecr.cr45p.s),
 Parameter(ecr.cr45p.w),
 Parameter(ecr.xp.amp),
 Parameter(ecr.xp.dur)}
```
Here, we see that `ecr.ch0.1` has a problem. What is the scope and what is the name. Indeed, I cannot get the parameter named `ch0.1` if I do:
```python
ecr2.get_parameters("ch0.1", scope="ecr")
```
instead I need to do:
```python
ecr2.get_parameters("1", scope="ecr.ch0")
```
which is not really intended. Here, I see two ways out:

1. forbid `.` in parameter names which is unfortunate since we chose that for channels in calibrations.
2. Use lists to manage scopes.

I would prefer option 2. Here we could do
```python
ecr2.get_parameters("amp", scope=("ecr", "cr45p"))
```
The only issue I see here is what parameter names do we use in `ecr2.scoped_parameters`? Or perhaps `ecr2.scoped_parameters` could return
```
{
 ("ecr", ): Parameter(ch0),
 ("ecr", ): Parameter(ch0.1),
 ("ecr", "cr45p"): Parameter(amp),
 ("ecr", "xp"): Parameter(amp),
 ...
}
```

eggerdj commented on qiskit/pulse/schedule.py:
```diff
@@ -814,69 +819,176 @@ def wrapper(self, *args, **kwargs):
+    Several preset contexts are available in :ref:`pulse_alignments`.
+    Schedule block is instantiated with one of above alignment contexts.
+    Default context is :class:`AlignLeft`, in which all instructions are left-justified,
+    in other word, as-soon-as-possible scheduling.
```
```suggestion
    in other words, as-soon-as-possible scheduling.
```

eggerdj commented on qiskit/pulse/schedule.py:
```diff
@@ -814,69 +819,176 @@ def wrapper(self, *args, **kwargs):
+    Default context is :class:`AlignLeft`, in which all instructions are left-justified,
+    in other word, as-soon-as-possible scheduling.
+
+    If you need absolute-time interval in between instructions, you can explicitly
+    insert :class:`~qiskit.pulse.instructions.Delay` instruction.
```
```suggestion
    If you need an absolute-time interval in between instructions, you can explicitly
    insert :class:`~qiskit.pulse.instructions.Delay` instructions.
```

eggerdj commented on qiskit/pulse/schedule.py:
```diff
@@ -814,69 +819,176 @@ def wrapper(self, *args, **kwargs):
+    .. rubric:: Nested blocks
+
+    Schedule block can nest other blocks with different alignment context.
+    This enables complicated scheduling, where a subset of instructions is
```
```suggestion
    This enables advanced scheduling, where a subset of instructions is
```

eggerdj commented on qiskit/pulse/schedule.py:
```diff
@@ -814,69 +819,176 @@ def wrapper(self, *args, **kwargs):
+    This enables complicated scheduling, where a subset of instructions is
+    locally scheduled in different manner.
+    Note that :class:`Schedule` instance cannot be directly added to the schedule block.
+    Schedule instances should be wrapped by :class:`Call` instruction to be added.
```
```suggestion
    To add Schedule instances wrap in a :class:`Call` instruction.
```

eggerdj commented on qiskit/pulse/schedule.py:
```diff
@@ -814,69 +819,176 @@ def wrapper(self, *args, **kwargs):
+    locally scheduled in different manner.
+    Note that :class:`Schedule` instance cannot be directly added to the schedule block.
+    Schedule instances should be wrapped by :class:`Call` instruction to be added.
+    This is implicitly performed when a schedule added through the :ref:`pulse_builder`.
```
```suggestion
    This is implicitly performed when a schedule is added through the :ref:`pulse_builder`.
```

eggerdj commented on qiskit/pulse/schedule.py:
```diff
@@ -814,69 +819,176 @@ def wrapper(self, *args, **kwargs):
+
+    .. rubric:: Unsupported operations
+
+    Because the schedule block is the representation lacking timeslots, it cannot
```
```suggestion
    Because the schedule block representation lacks timeslots, it cannot
```

eggerdj commented on qiskit/pulse/schedule.py:
```diff
@@ -814,69 +819,176 @@ def wrapper(self, *args, **kwargs):
+    .. rubric:: Subroutine
+
+    Timeslots-free representation offers much higher flexibility to write pulse programs.
+    Because it only cares the ordering of the child blocks,
```
```suggestion
    Because it only cares about the ordering of the child blocks,
```

eggerdj commented on qiskit/pulse/schedule.py:
```diff
@@ -814,69 +819,176 @@ def wrapper(self, *args, **kwargs):
+    Such subroutine is realized by the special compiler directive
+    :class:`~qiskit.pulse.instructions.Reference` that is defined by
+    a unique reference key to the subroutine and associated channels.
+    The substantial (executable) subroutine is separately stored in the main program.
```
```suggestion
    The (executable) subroutine is separately stored in the main program.
```

eggerdj commented on qiskit/pulse/schedule.py:
```diff
@@ -814,69 +819,176 @@ def wrapper(self, *args, **kwargs):
+
+    When you call a subroutine from another subroutine, or append a schedule block
+    to another schedule block, the management of references and parameters
+    may become hard task. Schedule block offers a convenient feature to help this
```
```suggestion
    can be a hard task. Schedule block offers a convenient feature to help with this
```

eggerdj commented on qiskit/pulse/schedule.py:
```diff
@@ -814,69 +819,176 @@ def wrapper(self, *args, **kwargs):
-      equal interval within a specified duration. Instructions on different channels
-      are aligned sequentially.
+    You may want to call this program from another program.
+    Here the program is called with the reference key ``grand_child``.
```
```suggestion
    In this example, the program is called with the reference key ``grand_child``.
```
?

eggerdj commented on qiskit/pulse/schedule.py:
```diff
@@ -1025,12 +1147,51 @@ def instructions(self) -> Tuple[Tuple[int, Instruction]]:
-        return tuple(self._blocks)
+        return tuple(self._blocks.nodes())
+
+    def _collect_parameters(self, scope: str, add_scope: bool = False) -> Set[Parameter]:
```
I wonder if `scope` should be `scope: Tuple[str, ...]` instead to avoid reserving `.` as a special character.

eggerdj commented on qiskit/pulse/schedule.py:
```diff
@@ -1222,24 +1402,154 @@ def assign_parameters(
+        subroutine_dict: Dict[str, "ScheduleBlock"],
+        inplace: bool = True,
+    ) -> "ScheduleBlock":
+        """Assign schedule to reference.
```
```suggestion
        """Assign schedules to references.
```

eggerdj commented on qiskit/pulse/schedule.py:
```diff
@@ -1222,24 +1402,154 @@ def assign_parameters(
+    ) -> "ScheduleBlock":
+        """Assign schedule to reference.
+
+        It is only capable of assigning a substantial program to child subroutines
```
I don't understand what is meant by `substantial program`?

nkanazawa1989 replied:
replaced with schedule block. "substantial" is used as an opposite of "opaque" because reference is definition-less instruction.

eggerdj commented on qiskit/pulse/schedule.py:
```diff
@@ -1222,24 +1402,154 @@ def assign_parameters(
+            with pulse.build(name="C") as main_prog:
+                pulse.call(name="B", channels=[pulse.DriveChannel(0)])
+
+        In above example, the ``main_prog`` ("C") is only aware of the subroutine "C.B" and its
+        reference of "B.A" is not exposed to the namespace of "C".
```
```suggestion
        In above example, the ``main_prog`` ("C") is only aware of the subroutine "C.B" and the
        reference of "B" to program "A", i.e., "B.A", is not exposed to the namespace of "C".
```

eggerdj commented on qiskit/pulse/schedule.py:
```diff
@@ -1222,24 +1402,154 @@ def assign_parameters(
+
+        In above example, the ``main_prog`` ("C") is only aware of the subroutine "C.B" and its
+        reference of "B.A" is not exposed to the namespace of "C".
+        This is because to prevent the breaking reference by the assignment of "C.B".
```
```suggestion
        This prevents breaking the reference ??? by the assignment of "C.B".
```
I'm not sure what reference is being mentioned here. Can you replace the `???` with the correct program?

eggerdj commented on qiskit/pulse/schedule.py:
```diff
@@ -1222,24 +1402,154 @@ def assign_parameters(
+    def get_parameters(
+        self,
+        parameter_name: str,
+        scope: Optional[str] = None,
```
```suggestion
        scope: Optional[Tuple[str, ...]] = None,
```
?

eggerdj commented on qiskit/pulse/schedule.py:
```diff
@@ -1222,24 +1402,154 @@ def assign_parameters(
-        Because different ``Parameter`` objects can have the same name,
-        this method returns a list of ``Parameter`` s for the provided name.
+        Note that we can define different parameter objects with the same name,
+        because these object is identified by the object uuid.
```
```suggestion
        because these different objects are identified by their unique uuid.
```

eggerdj commented on test/python/pulse/test_builder.py:
```diff
@@ -1229,3 +1229,324 @@ def test_call_subroutine_with_parametrized_duration(self):
         self.assertEqual(len(main.blocks), 1)
+
+
+class TestSubroutineWithCXGate(QiskitTestCase):
```
I think this test should cover the case where the name of a parameter has a `.` in it.

nkanazawa1989 replied:
Done in https://github.com/Qiskit/qiskit-terra/pull/8005/commits/b4a90adf082b691dc67851d2837c2b45d6ea0953

wshanks reviewed (commented):
Here is some initial review.

I want to think a bit more about some of the big picture of what is going on here -- how the scoping system works and how it is likely to be used in practice. The changes here do a lot more than the placeholder system I had originally been thinking of (not a bad thing).

One thing I would still like to look at more is the requirement that the reference include channels. The use case I envisioned for a referenced schedule would work better without needing to specify channels. To be a full Schedule/ScheduleBlock channels are needed but for the early specification of a schedule block calling another schedule block I don't think they are. With the way you make the central data structure a DAG here they become necessary but when ScheduleBlock was fundamentally a list of blocks I think not having channels could work (some things would not work but you wouldn't want those things until you filled in the reference with something with channels which typically would happen very early in the lifecycle of a ScheduleBlock with a reference).

nkanazawa1989 commented on qiskit/pulse/builder.py:
```diff
@@ -1892,90 +1917,136 @@ def snapshot(label: str, snapshot_type: str = "statevector"):
-            with string parameter names. If there are parameter name overlapping,
-            these parameters are updated with the same assigned value.
+        channels: Optional. Channels associated to the subroutine.
+        value_dict: Optional. Local scoped parameters assigned to the subroutine.
```
Thanks, this is outdated. Previously parameters are not assigned to the nested references. Now `assign_parameters` recursively assigns parameters to nested references.

nkanazawa1989 commented on qiskit/pulse/instructions/delay.py:
```diff
@@ -67,7 +67,3 @@ def channels(self) -> Tuple[Channel]:
         """Duration of this instruction."""
         return self.operands[0]
-
-    def is_parameterized(self) -> bool:
```
This is because base class method (`Instruction.is_parameterized`) is updated. Since all pulse instructions take the from of `Class(*operands)`, we don't need to write a dedicated logic to investigate operands for each class. We just need to write a base class logic that investigates parameters of `self._operands`.

nkanazawa1989 commented on qiskit/pulse/builder.py:
```diff
@@ -773,39 +790,47 @@ def call_subroutine(
                 "Schedule, or ScheduleBlock."
             )
 
-        if not empty_subroutine:
-            param_value_map = {}
-            for param_name, assigned_value in kw_params.items():
```
The change is because now `Call` instruction only takes `Schedule`. Since we have the reference mechanism for schedule block, we can create a reference instruction and assign the ``target`` immediately when schedule block is called as a target. (I lean toward deprecating Call in future, because of its complicated implementation which is hard to QPY serialize).

before:

- target is `Schedule` -> Wrap in `Call` instruction then append
- target is `ScheduleBlock` -> Wrap in `Call` instruction then append
- target is `QuantumCircuit` -> call schedule to create `Schedule` then wrap it in `Call` and append

now:

- target is `Schedule` -> same
- target is `QuantumCircuit` -> same
- target is `ScheduleBlock` -> create `Reference` and append, then call `assign_references` immediately with target
- target is `str` -> create `Reference`

nkanazawa1989 commented on qiskit/pulse/builder.py:
```diff
@@ -773,39 +790,47 @@ def call_subroutine(
+        sub_name = name or subroutine.name
+
+        if isinstance(subroutine, ScheduleBlock):
+            # If subroutine is schedule block, use reference mechanism.
```
I'm thinking to remove the use case of calling block as a `Call` instruction. I don't think there are many users directly building a pulse program without builder syntax, so this will be kind of internal mechanism update. 

The reason `Call` is introduced is to improve memory efficiency at program construction and serialization. A `Call` instance separately stores parameter-unassigned schedule and parameter table, and it is intended to reuse unassigned schedule among multiple instances. However, it is difficult to perform deduplication of schedules stored in separate instruction instances, on the other hand, reference mechanism easily realizes this because subroutine is now stored in the schedule block. 
Reference mechanism doesn't separately keep parameter-unassigned schedule, but I don't have any use case in mind that requires assignment of different parameters to the same parameterized sequence within the same schedule (i.e. Rabi experiment scans pulse amp, but each scan is different schedule instance). Indeed, current `Call` just degrades memory efficiency by having two schedule instances (one unassigned and other assigned one in cache) within is instance.

If we keep calling the schedule block, then called schedule is not recognized as a reference. If a program has mixed representation of call and reference, it seems like we can easily run into some edge case.

nkanazawa1989 commented on qiskit/pulse/schedule.py:
```diff
@@ -1025,12 +1147,51 @@ def instructions(self) -> Tuple[Tuple[int, Instruction]]:
     @property
     def blocks(self) -> Tuple["BlockComponent", ...]:
         """Get the time-ordered instructions from self."""
```
If I understand this correctly this is not time-ordered, but this is in order of addition. Now `self._block` is a retworkx graph connected with edges representing time ordering (i.e. parallel and sequential nodes), but if we get nodes with `self._block.nodes()` I think this doesn't care about edges. Correct? @mtreinish 

Anyways instruction is scheduled in order of addition (like FIFO), so it is close to time-ordering. Regarding the parallel context, this doesn't need to be time ordered, for example,

```python
with pulse.build() as sched:
    with pulse.align_right():
        pulse.play(pulse.Constant(100, 0.1, name="pulse1"), pulse.DriveChannel(0))
        pulse.play(pulse.Constant(200, 0.1, name="pulse2"), pulse.DriveChannel(1))
```

here the pulse1 starts at t=100 while pulse2 at t=0, but pulse1 is the first element of the returned list.

nkanazawa1989 replied:
I should update the documentation. Done in https://github.com/Qiskit/qiskit-terra/pull/8005/commits/372b684644a295f4c2ea9422bd3d625ebe67e324

nkanazawa1989 commented:
Thanks @eggerdj @wshanks for reviewing. I tend to agree with Will about channel-less representation of reference. We can define something like `OpaqueChannel` for DAG evaluation. IIRC, the requirements for channel is based on Daniel's request. Having channel in reference makes cal data structure cleaner? @eggerdj 

Regarding separator, it still prefer a single string rather than tuple of it. This makes serialization much easier. Also we can use regular expression for parameter search, for example, `scope=\S.x1` will give you parameters defined within the scope `x1` and its parent doesn't matter. I think this somewhat convenient. 

Perhaps we can conform to [SCPI](https://en.wikipedia.org/wiki/Standard_Commands_for_Programmable_Instruments)-ish syntax? For example, `x1:y1:amp` rather than `x1.y1.amp`. This allows you to use `.` in parameter name.

wshanks reviewed (commented):
This is looking good. I flagged some points to be discussed a little more.

eggerdj commented on qiskit/pulse/instructions/reference.py:
```diff
@@ -0,0 +1,91 @@
+            name: Optional. A label of this reference.
+
+        Raises:
+            PulseError: When no key is provided.
```
```suggestion
            PulseError: When no key is provided, i.e. ``ref_keys`` has length 0.
```

eggerdj commented on qiskit/pulse/instructions/reference.py:
```diff
@@ -0,0 +1,91 @@
+
+        Raises:
+            PulseError: When no key is provided.
+            PulseError: When key is not a string.
```
```suggestion
            PulseError: When a key is not a string.
```

eggerdj commented on qiskit/pulse/instructions/reference.py:
```diff
@@ -0,0 +1,91 @@
+        Raises:
+            PulseError: When no key is provided.
+            PulseError: When key is not a string.
+            PulseError: When key contains scope delimiter.
```
```suggestion
            PulseError: When a key in ``ref_keys`` contains the scope delimiter.
```

eggerdj commented on qiskit/pulse/instructions/reference.py:
```diff
@@ -0,0 +1,91 @@
+        """Create new reference.
+
+        Args:
+            ref_keys: A set of string that represent this reference.
```
This needs a bit more information. It is not clear from the doc string why ``ref_keys`` can be a set of strings and not just a string.

eggerdj commented on qiskit/pulse/reference_manager.py:
```diff
@@ -0,0 +1,44 @@
+    """Dictionary wrapper to manage pulse schedule references."""
+
+    def unassigned(self) -> Tuple[Tuple[str, ...], ...]:
+        """Get unassigned reference keys.
```
Minor
```suggestion
        """Get the keys of unassigned references.
```

eggerdj commented on qiskit/pulse/schedule.py:
```diff
@@ -814,69 +819,186 @@ def wrapper(self, *args, **kwargs):
+        print(main.scoped_parameters)
+
+    This implicitly creates a reference within the root program and assign ``sched2`` to it.
+    You get three parameters "root.amp", "root.child.amp", and "root.child.grand_child.amp".
```
This needs to be adjusted to the delimiter, i.e. `:`, no? Or perhaps `::` as @wshanks suggests.

nkanazawa1989 replied:
Yes, thanks for catching this :)

eggerdj commented on qiskit/pulse/schedule.py:
```diff
@@ -1011,9 +1140,13 @@ def duration(self) -> int:
     def channels(self) -> Tuple[Channel]:
         """Returns channels that this schedule clock uses."""
```
```suggestion
        """Returns channels that this schedule block uses."""
```

eggerdj commented on qiskit/pulse/instructions/reference.py:
```diff
@@ -0,0 +1,91 @@
+    """
+
+    # Delimiter for representing nested scope.
+    scope_delimiter = ":"
```
I like the suggestion of using `::` as a delimiter. This mirrors the way cpp defines namespaces, e.g. `std::cout`, so it has some rational. I also agree that the eye breaks harder on `,` which suggests `,` as a stronger scope but that is not really the intent. IIRC we said that the ``ref_keys`` would typically be sth of the form `(name of referenced schedule, channel A, channel B, ...)` (or was it qubits?). I wonder if we should make the intent of this structure more visible? E.g. `root::y1(d0)::x1(d0)::amp`? @nkanazawa1989 and @wshanks ? 

Also the example that Will shows above makes me wonder if there should be any checking in nested references. For example, I interpret `root::y1(d0)::x1(d0)::amp` as `y1(d0)` is a reference which plays on `d0` and contains itself a reference that plays `x1` on `d0`. Therefore, how do I interpret `root::y1(d0)::x1(d1)::amp`? Should this be allowed?

eggerdj commented on qiskit/pulse/builder.py:
```diff
@@ -1892,89 +1925,159 @@ def snapshot(label: str, snapshot_type: str = "statevector"):
+        from qiskit import pulse
+
+        with pulse.build() as main_prog:
+            pulse.refer("x_gate", "q0")
```
Should this be
```suggestion
            pulse.refer(("x_gate", "q0"))
```
?

nkanazawa1989 replied:
This is variadic argument, so not needed be a tuple here (it is internally converted into tuple).

eggerdj commented on qiskit/pulse/instructions/reference.py:
```diff
@@ -0,0 +1,91 @@
+    # Delimiter for representing nested scope.
+    scope_delimiter = ":"
+
+    def __init__(self, *ref_keys: str, name: Optional[str] = None):
```
Would `label` be more appropriate than name? I.e. something that could be used to indicate the intent of the reference. As example, for a wordy user `label="dd_xpulse"` vs `label="an_algorithm_needed_xpulse"` to indicate which of his X-pulses are for dynamical decoupling and which ones are needed by his quantum algorithm.

nkanazawa1989 replied:
This makes sense. `name` is kind of required argument of the pulse instruction superclass. It's straightforward to use `name` as a subroutine name, and turn `*ref_keys` into `*extra_keys`. For example, ("ecr", "q0", "q1") is interpreted as subroutine name="ecr" with extra Information ("q0", "q1"). The net behavior doesn't change, but we can simplify some validation. See https://github.com/Qiskit/qiskit-terra/pull/8005/commits/6d874108fcebbdb687d9a58f8514e557bea40684

nkanazawa1989 replied:
I think label can be added later if necessary.

eggerdj commented on qiskit/pulse/builder.py:
```diff
@@ -740,6 +744,16 @@ def append_block(self, context_block: ScheduleBlock):
+            ref_keys: Set of keys to uniquely specify the reference.
+            name: Optional. A label of instruction.
+        """
+        reference = instructions.Reference(*ref_keys, name=name)
+        self.append_instruction(reference)
```
Here I have a question: Should this be doing any validation? This could be needed for nested references, e.g.,`root::y1(d0)::x1(d0)` seems to make sense (referenced `y1` on channel `d0` references `x1` played on channel `d0`, which is consistent) but `root::y1(d0)::x1(d1)` does not (referenced `y1` on channel `d0` references `x1` played on channel `d1`, which is not consistent).

nkanazawa1989 replied:
As @wshanks suggested, reference is not aware of channels. This allows us to avoid complicated assignment problems, i.e. channel or qubit might be parameterized. So channels of the reference is treated unknown until assignment. But you can provide some information about channel or qubit in key tuples on your responsibility, i.e. that management is delegated to calibrations.

nkanazawa1989 commented on qiskit/pulse/schedule.py:
```diff
@@ -901,80 +901,90 @@ class ScheduleBlock:
+
+    .. jupyter-execute::
+
+        main.get_parameters("amp", scope="\\S.grand_child")
```
Good catch! thanks

nkanazawa1989 commented on qiskit/pulse/schedule.py:
```diff
@@ -1222,24 +1464,166 @@ def assign_parameters(
         Args:
             parameter_name: Name of parameter.
+            scope: Scope of the program to investigate. Defaults to the current block.
+                Scope can be specified with regular expression.
```
Make sense

nkanazawa1989 commented on qiskit/pulse/utils.py:
```diff
@@ -114,3 +115,12 @@ def wrapper(*args, **kwargs):
     return wrapper
+
+
+def scope_parameter(param: Parameter, scope: str, delimiter: str = ":") -> Parameter:
```
I just moved it to schedule as a protected. `rename_parameter` is very general and we may want to put this in `qiskit.circuit` where parameter object is defined. But this is currently only used by schedule as you mentioned.

Done in https://github.com/Qiskit/qiskit-terra/pull/8005/commits/944076c033ac03a385145ae110e7c8aef2f80ea7

nkanazawa1989 commented on qiskit/pulse/schedule.py:
```diff
@@ -1222,24 +1464,166 @@ def assign_parameters(
+
+        matched = set()
+        for param in self.scoped_parameters:
+            scope_and_name = param.name.split(Reference.scope_delimiter)
```
This is good to know. Thanks! Done in https://github.com/Qiskit/qiskit-terra/pull/8005/commits/944076c033ac03a385145ae110e7c8aef2f80ea7

nkanazawa1989 commented on qiskit/pulse/builder.py:
```diff
@@ -814,22 +811,33 @@ def call_subroutine(
+                # https://docs.python.org/3.9/library/functions.html#id
+                sched_repr = repr(subroutine)
+                sched_params = sorted(subroutine.parameters, key=lambda p: p.name)
+                unique_str = hex(hash((sched_repr, *sched_params)))
```
Object identity might not work for deep copied object without returning from the scope. I agree this is delicate, and I removed the logic and simply added uuid instead.

nkanazawa1989 commented on qiskit/pulse/builder.py:
```diff
@@ -1984,70 +2026,58 @@ def call(
+    )
+
+
+def refer(*ref_keys: str, name: Optional[str] = None):
```
Done in https://github.com/Qiskit/qiskit-terra/pull/8005/commits/021577e9c33395f1870c1514e8a1adce3c7d2795

nkanazawa1989 commented on qiskit/pulse/transforms/dag.py:
```diff
@@ -59,9 +54,51 @@ def block_to_dag(block: "ScheduleBlock") -> rx.PyDAG:
+                slots[chan] = node_id
+        else:
+            for chan in elm.channels:
+                prev_id = slots.pop(chan, None)
```
Good catch! Updated in https://github.com/Qiskit/qiskit-terra/pull/8005/commits/44d4959da2e1022f56f2815487f597626f6f80ee I also found another edge case of nested schedule block with unassisted reference. Test is also added.

nkanazawa1989 commented on qiskit/pulse/schedule.py:
```diff
@@ -1133,9 +1140,13 @@ def duration(self) -> int:
-                chans.add(chan)
+        for elm in self.blocks:
+            if isinstance(elm, Reference):
+                raise PulseError(
```
This should be okey, i.e. `self.block` is property that returns blocks after assignment, i.e. self._blocks is raw sequence including reference instruction.

nkanazawa1989 commented on qiskit/pulse/instructions/reference.py:
```diff
@@ -0,0 +1,91 @@
+        for key in ref_keys:
+            if not isinstance(key, str):
+                raise PulseError(f"Keys must be string. '{repr(key)}' is not a valid object.")
+            if self.scope_delimiter in key:
```
`search_parameters` sounds like very good idea. Done in https://github.com/Qiskit/qiskit-terra/pull/8005/commits/f92fa65a5f11b9e248d31c0c652c1f5773d35f3d Also use of key delimiter is explicitly checked in `Reference` instruction constructor.

nkanazawa1989 commented on qiskit/pulse/instructions/call.py:
```diff
@@ -35,7 +32,7 @@ class Call(instruction.Instruction):
     def __init__(
         self,
-        subroutine: "Schedule",
+        subroutine,
```
I just reverted the change. This is to avoid cyclic import.

wshanks commented on test/python/pulse/test_reference.py:
```diff
@@ -0,0 +1,633 @@
+        """Test call and append schedule.
+
+        Reference is copied to the outer schedule by appending.
+        Original reference is remain unchanged.
```
```suggestion
        Original reference remains unchanged.
```

wshanks commented on test/python/pulse/test_reference.py:
```diff
@@ -0,0 +1,633 @@
+        User can append blocks which is an element of another schedule block.
+        But this is not standard usecase.
+
+        In this case, references may contain subroutines which don't exit in the context.
```
```suggestion
        In this case, references may contain subroutines which don't exist in the context.
```

wshanks commented on test/python/pulse/test_reference.py:
```diff
@@ -0,0 +1,633 @@
+        But this is not standard usecase.
+
+        In this case, references may contain subroutines which don't exit in the context.
+        This is because all references within the program is centralizedly
```
```suggestion
        This is because all references within the program are centrally
```

wshanks commented on test/python/pulse/test_reference.py:
```diff
@@ -0,0 +1,633 @@
+
+        with pulse.build() as sched_z2:
+            # Append child block
+            # The refecence of this block is sched_z1.reference thus it contains both x1 and y1.
```
```suggestion
            # The reference of this block is sched_z1.reference thus it contains both x1 and y1.
```

wshanks commented on qiskit/pulse/schedule.py:
```diff
@@ -1066,15 +1267,40 @@ def append(
+                    # Move only references existing in the 'block'.
+                    # See 'test.python.pulse.test_reference.TestReference.test_appending_child_block'
+                    for ref in _get_references(block._blocks):
+                        self.references[ref.ref_keys] = block.references[ref.ref_keys]
```
What if one of these references is already assigned in self?

nkanazawa1989 replied:
Thanks, done in https://github.com/Qiskit/qiskit-terra/pull/8005/commits/670b333f500c276cd174dcfe49f5179f153c7e3c

wshanks reviewed (commented):
This is mostly a bunch of wording suggestions for the docs. I am pretty happy with the implementation. I had a few small other comments.

One thing I wonder about is the sense of scoping for references and parameters. It seems like root schedule blocks deal with the parameters of all descendant references but only with immediate reference keys. You explain in the documentation that this is because tracking grandchildren in the root block would get complicated if the child's references were also modified directly. I don't think the behavior needs to change but we might want to make sure the language is consistent. Some of it talks broadly about a block only knowing about its immediate children but that is not the case for parameters.

nkanazawa1989 commented on qiskit/pulse/reference_manager.py:
```diff
@@ -0,0 +1,44 @@
+        out = f"{self.__class__.__name__}:"
+        for key, reference in self.items():
+            prog_repr = repr(reference)
+            if len(prog_repr) > 50:
```
No it generates a string representation of the full schedule which is usually quite hard to read. We should implement pretty print of schedule block in followup.

nkanazawa1989 commented on qiskit/pulse/schedule.py:
```diff
@@ -1024,13 +1160,78 @@ def instructions(self) -> Tuple[Tuple[int, Instruction]]:
+            add_scope: Set ``True`` to prepend scope to parameter name.
+
+        Returns:
+            All parameters defined under the current scope.
```
Oh, good catch. Updated the logic to keep (same) parameter object in the different scope in scope-aware mode. https://github.com/Qiskit/qiskit-terra/pull/8005/commits/f24745a762dfe3a1c86b70a8e305a1dbf814403f

nkanazawa1989 commented on qiskit/pulse/schedule.py:
```diff
@@ -1222,24 +1467,169 @@ def assign_parameters(
         """
-        return self._parameter_manager.get_parameters(parameter_name)
+        matched = [p for p in self.parameters if p.name == parameter_name]
+        return sorted(matched, key=lambda p: p.name)
```
Good catch. I just copied it from another method, but sorting is unnecessary here. Fixed in https://github.com/Qiskit/qiskit-terra/pull/8005/commits/f24745a762dfe3a1c86b70a8e305a1dbf814403f

nkanazawa1989 commented on qiskit/pulse/schedule.py:
```diff
@@ -1066,15 +1267,40 @@ def append(
+            return schedule
+
+        if isinstance(block, Reference):
+            self.references[block.ref_keys] = None
```
Thanks, that's an edge case. Updated the logic in https://github.com/Qiskit/qiskit-terra/pull/8005/commits/670b333f500c276cd174dcfe49f5179f153c7e3c

nkanazawa1989 commented on qiskit/pulse/schedule.py:
```diff
@@ -1066,15 +1267,40 @@ def append(
+                    self.references.update(block._reference_manager)
+                    block._reference_manager.clear()
+            # Now switch the parent because block is appended to self.
+            block._parent = self
```
No, this is due to nesting feature of schedule blocks. For example,

```python
with build() as root:
    with align_left():
        with align_right():
            reference("sub")
```
In this example, we should be able to assign subroutine with
```python
root.assign_references({("sub", ): subroutine})
```
because every reference in the same builder context should be exposed to `root`, i.e.  nesting a block doesn't create reference. However, because of current implementation bottleneck, ref manager is attached to each block, and thus, without parent referencing mechanism, we must write this awkward code:
```python
root.blocks[0].blocks[0].assign_references({("sub", ): subroutine})
```
The parent reference allows us to manage references defined in all nested blocks in the most outer block.

wshanks replied:
Okay, that makes sense. Maybe a comment on `_parent` in `__init__` would be good. One thing I wonder about is substituted references. If `_parent` is only ever used for looking up the reference manager, there should be no problem since `.blocks` substitutes in the references and so references should not need to be considered further. When you call `.blocks` you will get a list of blocks with `_parent` pointing at the parent, except for the assigned references which will be `ScheduleBlock`s with `_parent == None`. So I could see potential confusion in the future if `_parent` were used in some other way.

nkanazawa1989 replied:
Thanks, Done in https://github.com/Qiskit/qiskit-terra/pull/8005/commits/620e67d62b584537cbd05c741bbeea0dab80fe6a

nkanazawa1989 commented:
Thanks @wshanks for very careful review. I found two edge cases from your questions, namely, unexpected merging of scoped parameter (https://github.com/Qiskit/qiskit-terra/pull/8005/commits/f24745a762dfe3a1c86b70a8e305a1dbf814403f) and overriding existing reference (https://github.com/Qiskit/qiskit-terra/pull/8005/commits/670b333f500c276cd174dcfe49f5179f153c7e3c). Test has been updated to cover these situations. Documentation has been updated too.

wshanks commented on qiskit/pulse/schedule.py:
```diff
@@ -1222,24 +1490,173 @@ def assign_parameters(
+
+        .. note::
+
+            Assigned programs are deep-copied to prevent an unexpected update.
```
What is the concern here?

nkanazawa1989 replied:
This is to make the principle clear. At the level of schedule, a key and referenced object should be unique mapping. So even if the source is the same object, it will be independent when assigned to the different keys. Thus, parameter link (mapto) should be implemented at the level of the calibrations framework.

wshanks reviewed (approved):
It looks good to me now. I am a little concerned about some of the intricacy of managing parameters and references but we can see how it goes. I think for the most part this will just be used for one level of referencing without mixed parameters between the parent and child which is a case that is not so complex.

I put a few small questions about some of the subtle parts (copying).

We should take a look at @eggerdj's CR library implementation before merging to double check that this fits that use case, since that was the original motivation for this.

eggerdj commented:
This looks good to me. See e.g. https://github.com/Qiskit/qiskit-experiments/pull/889 showing how it is used.

nkanazawa1989 commented on qiskit/pulse/schedule.py:
```diff
@@ -1066,15 +1286,41 @@ def append(
+            self.references[block.ref_keys] = None
+
+        elif isinstance(block, ScheduleBlock):
+            block = copy.deepcopy(block)
```
Basically appended schedule will lose it's reference manager data and parent schedule is set to the current scope. So this is not to mutate the original by the system.

nkanazawa1989 commented on qiskit/pulse/schedule.py:
```diff
@@ -1066,15 +1286,41 @@ def append(
+            # The block is just directly appended to the current scope.
+            if block.is_referenced():
+                if block._parent is not None:
+                    # This is an edge case:
```
No, I think such usage should be prohibited, but currently I don't have good mechanism to do that.

jakelishman commented on qiskit/pulse/builder.py:
```diff
@@ -1832,89 +1854,165 @@ def snapshot(label: str, snapshot_type: str = "statevector"):
 
-        with pulse.build() as main_prog:
-            pulse.call(subroutine, value_dict={amp1: 0.1, amp2: 0.2})
+        If there is a name collision between parameters, you can distinguish them by specifying
+        each parameter object in a python dictionary. For example,
+
+        .. jupyter-execute::
+
+            amp1 = circuit.Parameter('amp')
+            amp2 = circuit.Parameter('amp')
+
+            with pulse.build() as subroutine:
+                pulse.play(pulse.Gaussian(160, amp1, 40), pulse.DriveChannel(0))
+                pulse.play(pulse.Gaussian(160, amp2, 40), pulse.DriveChannel(1))
+
+            with pulse.build() as pulse_prog:
+                pulse.call(subroutine, value_dict={amp1: 0.1, amp2: 0.3})
```
Is there a reason to allow the user to write a subroutine with a naming clash between its parameters?  Seems a bit odd.

wshanks replied:
I can't think of why you would set out to write a schedule like the example, but I could see it coming up in more complex code, like calling different functions to generate different parts of a schedule or using references to different schedules and then using `inline_subroutines` to merge them into one big schedule block.

jakelishman replied:
That's totally fine by me, if there's a reason.  I really don't know any of the context - I was basically just reviewing by "smell".

nkanazawa1989 replied:
Yeah basically this kind of problem occurs because the pulse module uses `Parameter` object designed for circuit as-is. Usually pulse parameter is scoped by pulse name (this allows user to safely have the same name in a schedule, i.e., sequence of a pulse), and thus UUID mechanism is not necessary. In pulse programming, it will never happen to have the same parameter name in the same pulse. So proper model should be `Parameter(scope, name)`, rather than `Parameter(name, uuid)`.

jakelishman commented on qiskit/pulse/instructions/instruction.py:
```diff
@@ -56,9 +57,12 @@ def __init__(
-        for channel in self.channels:
-            if not isinstance(channel, Channel):
-                raise PulseError(f"Expected a channel, got {channel} instead.")
+        try:
+            for channel in self.channels:
+                if not isinstance(channel, Channel):
+                    raise PulseError(f"Expected a channel, got {channel} instead.")
+        except UnassignedReferenceError:
+            pass
```
This _looks_ surprising to me, without knowing the context - what part of the `try` block can through an `UnassignedReferenceError`?  Nothing looks like it ought to be doing any computation, which suggests to me there's performance surprises hiding in the API.

wshanks replied:
This is a bit awkward. The issue is that the `Reference` instruction raises this error in its `channels` property. There is no reasonable value to give for `channels` for an unassigned reference, but on the other hand there is no good reason to try to look at the channels of a `ScheduleBlock` with unassigned references. The approach taken here is to raise on accessing `Reference.channels` and have `ScheduleBlock` substitute in the assigned subroutine (when it exists) in place of the `Reference` in its `channels` and `blocks` methods. So nothing outside of the `ScheduleBlock` should be working directly with a `Reference` instruction.

Perhaps there is a more elegant way to deal with this. For this specific change to `Instruction.__init__`, an alternative would be to have `Reference.channels` return an empty list until a flag is set. Then `Reference.__init__` could call `super().__init__` and then set this flag. `Reference.channels` could always return an empty list but I think raising an exception is better because any code trying to work with the channels of a `ScheduleBlock` with unassigned references is probably doing something wrong.

By the way, I don't know if it is clear reading the PR now, but the original motivation for the `Reference` instruction was to support defining a schedule like an echoed cross-resonance gate in a separate function (or even a separate package) from the one defining the single qubit x gate. The intent was that consumer would basically define the ecr and x gates and then assign x to ecr before doing anything else.

jakelishman replied:
Ah, so `channels` is a property that does a little bit of work?  I have no problems with raising the exception if there's nothing valid to return.

I personally don't like properties doing work (I much prefer having the explicit method call) because I find it's easy to introduce massive performance hits accidentally in consumer code (people assume attribute accesses are free) and if the return value is mutable, consumers often assume mutations will propagate back to the object.  But that's not my call to make, and if the work is only very slight anyway (and the return type is immutable), I wouldn't think it's a big deal.

(I honestly didn't even read the PR closely enough to know _what_ it was trying to add - I trust you and Daniel were better reviewers of that.)

wshanks replied:
The only work `channels` is doing at the moment is raising an exception, so it shouldn't have a performance impact. I am curious what @nkanazawa1989 thinks about having it return `[]` until after calling `super().__init__()` so that we don't have to have a special case for `Reference` in the base `Instrument.__init__` method.

nkanazawa1989 replied:
This is good idea but I don't have reasonable implementation, because `Instruction` doesn't have `channels` values. They are just a part of `self._operantds` and all subclasses report channel with known index (this is to reduce parameter assignment overhead). This means we need to dynamically rewrite method.

nkanazawa1989 replied:
I came up with another implementation in https://github.com/Qiskit/qiskit-terra/pull/8005/commits/cdefe1633a9eb392ae6c211bad1266af94098b52. I added explicit `_validate` method so that every subclass can override. Indeed there was redundant instance check in subclasses.

wshanks replied:
cdefe16 looks nice to me (but might be the source of the new test failure?).

nkanazawa1989 replied:
Oh I didn't notice the test was also here. Actually this test is a bug and the code here will be never accepted by real hardware. The OpenPulse simulator in Aer doesn't really simulate the measurement dynamics and thus this random channel is accepted by it.

jakelishman commented on qiskit/pulse/instructions/instruction.py:
```diff
@@ -197,16 +201,25 @@ def append(self, schedule, name: Optional[str] = None):
+                    if isinstance(param, Parameter):
+                        params.add(param)
+                    else:
+                        param = param | _get_parameters_recursive(param)
```
did you mean `params |= _get_parameters_recursive(param)`? (note the different assignment)

wshanks replied:
I think you are right and this probably means there needs to be another test added that uses a `ParameterExpression`.

nkanazawa1989 replied:
Thanks, updated in https://github.com/Qiskit/qiskit-terra/pull/8005/commits/329c6bc386e81c8a73f6da60e64d3bc47a492f9e

jakelishman commented on qiskit/pulse/parameter_manager.py:
```diff
@@ -50,7 +50,7 @@
 and thus this parameter framework gives greater scalability to the pulse module.
 """
-from copy import deepcopy, copy
+from copy import copy
```
I'm very glad to see one fewer use of `deepcopy` in our code base!

jakelishman commented on qiskit/pulse/parameter_manager.py:
```diff
@@ -300,18 +300,19 @@ def visit_ScheduleBlock(self, node: ScheduleBlock):
-            self.parameters.add(parameter)
+        # Note that node.parameters returns parameters of main program with subroutines.
+        # The manager of main program is not aware of parameters in subroutines.
+        self.parameters = self.parameters | node._parameter_manager.parameters
```
Very minor, but here and elsewhere, `self.parameters |= node._parameter_manager.parameters` is more performant (if it's safe) because it does in-place mutation without creating a new object.

nkanazawa1989 replied:
Thanks done in https://github.com/Qiskit/qiskit-terra/pull/8005/commits/329c6bc386e81c8a73f6da60e64d3bc47a492f9e

jakelishman commented on qiskit/pulse/parameter_manager.py:
```diff
@@ -300,18 +300,19 @@ def visit_ScheduleBlock(self, node: ScheduleBlock):
         """Get parameters from block's ``AlignmentKind`` specification."""
         for param in node._context_params:
-            self.visit(param)
+            if isinstance(param, ParameterExpression):
+                self.parameters = self.parameters | param.parameters
```
Is there any chance you might need to recurse into `visit(param)` if it's no a `ParameterExpression`?

nkanazawa1989 replied:
Not for now. User can still add custom alignment by subclassing `AlignmentKind` and such a custom alignment could take some object which is not `ParameterExpression`. But I think I've never come up with such fancy alignment logic.

jakelishman commented on qiskit/pulse/reference_manager.py:
```diff
@@ -0,0 +1,57 @@
+                keys.append(key)
+        return tuple(keys)
+
+    def __setitem__(self, key, value):
+        if key in self and self[key] is not None:
+            # Check subroutine conflict.
+            if self[key] != value:
+                raise PulseError(
+                    f"Subroutine {key} is already assigned to the reference of the current scope, "
+                    "however, the newly assigned schedule conflicts with the existing schedule. "
+                    "This operation was not successfully done."
+                )
+            return
+        super().__setitem__(key, value)
```
This isn't generally sufficient when subclassing `dict`:
```python
>>> class MyDict(dict):
...     def __setitem__(self, key, value):
...             print("here!")
...             return super().__setitem__(key, value)
...
>>> d = MyDict()
>>> d["hello"] = "world"
here!
>>> d.update({"a": "b"})
>>> # nothing printed!
```

The Python built-ins generally implement all their methods separately in C space, and don't call the Python-space overrides.  The quickest fix is just to derive from `collections.UserDict` instead, which will do what you expect.

nkanazawa1989 replied:
Thanks, done in https://github.com/Qiskit/qiskit-terra/pull/8005/commits/ae506e849ed3a93ce74a5c6f0ba910ee7c6fb2a7

jakelishman commented on qiskit/pulse/schedule.py:
```diff
@@ -956,9 +1094,15 @@ def initialize_from(cls, other_program: Any, name: Optional[str] = None) -> "Sch
+        """Return name of this schedule"""
         return self._name
 
+    @name.setter
+    def name(self, new_name: str):
+        """Update the name of this schedule."""
+        self._name = new_name
+        self._reference_manager._scope = new_name
+
```
Is it important for the name to become mutable?  That feels like it might have knock-on effects.

nkanazawa1989 replied:
No. This is some artifact from initial implementation. Removed in https://github.com/Qiskit/qiskit-terra/pull/8005/commits/f4834569b9961dd7761e0326925b5a318ed8c07c

jakelishman commented on qiskit/pulse/schedule.py:
```diff
@@ -1024,13 +1175,90 @@ def instructions(self) -> Tuple[Tuple[int, Instruction]]:
     @property
-    def parameters(self) -> Set:
-        """Parameters which determine the schedule behavior."""
-        return self._parameter_manager.parameters
+    def parameters(self) -> Tuple[Parameter]:
+        """Return unassigned parameters with raw names."""
+        return tuple(self._collect_parameters(scope="root", add_scope=False).values())
```
I think this is a public API change?  Is it absolutely necessary?

nkanazawa1989 replied:
No. API and behavior doesn't change. It additionally checks subroutines because parameters in there are explicitly unexposed to the current scope.

nkanazawa1989 replied:
Sorry you are right, this is unnecessary change. Fixed in https://github.com/Qiskit/qiskit-terra/pull/8005/commits/0fc47056ec17df4434ca476fdb2160f08124c7e8

jakelishman commented on qiskit/pulse/schedule.py:
```diff
@@ -1024,13 +1175,90 @@ def instructions(self) -> Tuple[Tuple[int, Instruction]]:
+        """Return unassigned parameters with raw names."""
+        return tuple(self._collect_parameters(scope="root", add_scope=False).values())
+
+    @property
+    def scoped_parameters(self) -> Tuple[Parameter]:
+        """Return unassigned parameters with scoped names.
+
+        .. note::
+
+            If a parameter is defined within a nested scope,
+            it is prefixed with all parent-scope names with the delimiter string,
+            which is "::". If a reference key of the scope consists of
+            multiple key strings, it will be represented by a single string joined with ",".
+            For example, "root::xgate,q0::amp" for the parameter "amp" defined in the
+            reference specified by the key strings ("xgate", "q0").
+        """
+        return tuple(self._collect_parameters(scope="root", add_scope=True).values())
```
I get nervous for performance reasons when properties do non-trivial work (similar to above, but there's an API concern there as well) - given that this is new, is it worth it being a method?

nkanazawa1989 replied:
Yes, I converted this into method with some refactoring https://github.com/Qiskit/qiskit-terra/pull/8005/commits/5b7eeef21deec5d01ad0fd542275f135d54d7c55

jakelishman commented on qiskit/pulse/schedule.py:
```diff
@@ -1180,38 +1434,60 @@ def replace(
+        self._parameter_manager.clear()
+        self._parameter_manager.update_parameter_table(self._alignment_context)
+        new_elms = []
+        for elm in self._blocks:
+            if elm == old:
+                elm = new
+            self._parameter_manager.update_parameter_table(elm)
+            new_elms.append(elm)
```
Can the same element be in the blocks more than once?  If so, it could get its parameter table updated twice which may cause an error?  If not, you can probably short-circuit the loop once you've seen the target to avoid what I assume are costly `==` comparisons.

nkanazawa1989 replied:
Yes, it can. Note that we cannot shortcut the loop because we need to refresh parameter manager, i.e. we could think of simply removing parameters associated with `old`, however, the same parameter could exist in other block elements. Thus we always need full scan. I think replace is not good method and usually it's better to write a schedule from scratch.

jakelishman replied:
Fine by me, as long as it was intentional.

jakelishman commented on qiskit/pulse/schedule.py:
```diff
@@ -1504,6 +1925,33 @@ def _get_timeslots(schedule: "ScheduleComponent") -> TimeSlots:
+    Returns:
+        A set of unique reference instructions.
+    """
+    references = []
+    for elm in block_elms:
+        if isinstance(elm, ScheduleBlock):
+            references += _get_references(elm._blocks)
+        elif isinstance(elm, Reference):
+            references.append(elm)
+    return set(references)
```
Seems odd not just to build up the set immediately?

nkanazawa1989 replied:
Good catch, definitely this is right way. Done in https://github.com/Qiskit/qiskit-terra/pull/8005/commits/7a19c58b09d7e3306295b7c2da85b89569429adc

jakelishman commented on qiskit/pulse/schedule.py:
```diff
@@ -1504,6 +1925,33 @@ def _get_timeslots(schedule: "ScheduleComponent") -> TimeSlots:
+    return set(references)
+
+
+def _scope_parameter(param: Parameter, scope: str, delimiter: str = ":") -> Parameter:
+    """Override parameter object with program scope information."""
+    new_name = f"{scope}{delimiter}{param.name}"
+    scoped_param = Parameter.__new__(Parameter, new_name, uuid=getattr(param, "_uuid"))
+    scoped_param.__init__(new_name)
+
+    return scoped_param
```
This looks a bit terrifying to me, to be honest.  I get what you're doing, but it feels potentially very fragile.  I don't have a sensible alternative suggestion, though, so it's probably best to leave it.

nkanazawa1989 replied:
This is identical trick used in qpy loader, but I agree this is bit hackey. I really wanted to introduce dedicated parameter subclass for pulse module.

jakelishman commented on qiskit/pulse/transforms/dag.py:
```diff
@@ -64,39 +64,44 @@ def block_to_dag(block: ScheduleBlock) -> rx.PyDAG:
-        prev_node = current_node
-    dag_blocks.add_edges_from_no_data(edges)
+    for elm in block.blocks:
+        node_id = dag.add_node(elm)
+        if dag.num_nodes() == 1:
+            continue
+        prev_id = dag.node_indices()[-1]
+        edges.append((prev_id, node_id))
```
I haven't tested it, but this _feels_ very like `prev_id` is just going to be the same as `node_id` every time?  If so, this isn't make a DAG - it's making a graph where every node points only to itself.

wshanks replied:
Hah, you are right. It should be making a DAG that is just a linear chain of all the elements but it is instead making a set of nodes pointing to themselves. I think this indicates a missing test -- if you make two schedule blocks with the same elements but in different orders, a test that checks that the two schedules are not equal should fail.

wshanks replied:
It turns out that there is good test coverage of the different equality cases in test_block.py. Here is the test I was suggesting:

https://github.com/Qiskit/qiskit-terra/blob/fe65c1a2cff8b1061425d92fb03ccc9534c5b1b4/test/python/pulse/test_block.py#L474-L484

However, there is a quirk that makes this test pass. The `if dag.num_nodes() == 1` condition causes the first node not to get an edge to itself. So for two-element schedules, the two elements end up unique -- the first one has no edges while the second one has an edge. This variant of the test fails for me though:

```python
        block1 = pulse.ScheduleBlock(alignment_context=self.sequential_context)
        block1 += pulse.Play(self.test_waveform0, self.d0)
        block1 += pulse.Play(self.test_waveform0, self.d0)
        block1 += pulse.Play(self.test_waveform0, self.d1)

        block2 = pulse.ScheduleBlock(alignment_context=self.sequential_context)
        block1 += pulse.Play(self.test_waveform0, self.d0)
        block2 += pulse.Play(self.test_waveform0, self.d1)
        block2 += pulse.Play(self.test_waveform0, self.d0)

        self.assertNotEqual(block1, block2)
```

In this case, the first element of each schedule is the same so that they compare okay as the no-edge parts of the graphs, and then the remaining elements all point to themselves. `is_isomorphic_node_match` can then reorder the remaining elements to match them since there are no edges between them.

jakelishman replied:
In fairness, your second Python block there has a `block1 +=` hidden among the `block2` mutations.  I don't know if that affects anything.

wshanks replied:
Oops, that was me redoing the modification in the comment instead of copy-pasting from the test file I modified. Fixed in the code above now.

nkanazawa1989 replied:
This can accidentally pass this test. Indeed this always creates closed edges, but this doesn't appear in the first node. So test can still pass up to two nodes. Code is fixed in https://github.com/Qiskit/qiskit-terra/pull/8005/commits/c973d61e2b2fa32cfec94e8340feb9ac2b89613e

Test from @wshanks looks effective to detect the bug. Added in https://github.com/Qiskit/qiskit-terra/pull/8005/commits/1531480175836a568ce1253ce77a328fca0bc983.

jakelishman reviewed (commented):
I tried to read through this large PR pretty quickly so it could move to merge, and especially since I'm not familiar with the pulse builders, I was generally just looking for stuff that _felt_ odd to me - please feel free to dismiss me if I'm wrong on it.

I'm largely trusting that if Will and Dan are happy with the interface as written, then it's good, my code-based comments notwithstanding.  I can't offer any meaningful opinions on that.

wshanks commented on qiskit/pulse/builder.py:
```diff
@@ -769,39 +784,46 @@ def call_subroutine(
+            # If subroutine is schedule block, use reference mechanism.
+            if local_assignment:
+                subroutine = subroutine.assign_parameters(local_assignment, inplace=False)
+            if name is None:
+                # Add unique string, not to accidentally override existing reference entry.
+                keys = (subroutine.name, uuid.uuid4().hex)
+            else:
+                keys = (name,)
```
I think always using shadowing is probably best. You need to use shadowing when the name is not given because a schedule could call multiple schedules with the same name (e.g. a two qubit schedule calling x on each qubit). While using the name as the key feels intuitive, it's not needed to preserve the way `call()` worked before.

One issue with the current behavior is that if you use the same name for multiple calls with different schedules (which does not seem like a good thing to do) the last one overwrites the earlier ones whereas before the individual schedules would be kept.

Besides backwards compatibility, the nice thing about `call()` is that adds a reference and assigns it in a single step. Maybe the new `reference()` method could optionally take a `ScheduleBlock` and assign it if given? Since it is a new method, it does not have to worry about backwards compatibility and can just always use the required `name` parameter as the key.

jakelishman replied:
I think we used opposite meanings of "shadowing" here - I meant "overwrites" (so the previous definitions of "x" are inaccessible now), and I think you meant "adds to"?  It doesn't matter too much if we did - I don't have any position on what's best for the API here, because I've never used the code.  As long as you and Naoki come to an agreement on what's best, I can sign off on the PR.  If that's the extra uuid bit of the key in all cases, that's absolutely fine by me.

wshanks replied:
Yes, you are right. I got it backwards from you. I was thinking of the function obscuring the subroutine's name as shadowing the name.

nkanazawa1989 replied:
This is good point. I'm not 100 percent confident on the current implementation. This at least guarantees the conventional behavior. The reference key is important only when a user assigns a subroutine to reference. Since, `.call` performs this operation in a single function, basically the reference key can be whatever. There is no reason to update explicitly provided `name`, but I added uuid here to avoid _local parameter collision_ since we can call `.call` with parameter assignment. Remember that reference manager does NOT allow assigning multiple objects to a single reference. Calling the same schedule with different parameter assignments will generate different subroutines with the same name on the fly, and this operation usually fails without shadowing. This should be supported for backward compatibility. I wanted to generate unique name considering parameters, but I needed to give up because of mutability of `.assign_parameters` (I needed unnecessarily heavy logic to realize). Another option would be 
```suggestion
            keys = name or uuid.uuid4().hex
```
probably this makes more sense since `subroutine.name` is no longer the effective identifier due to uuid.

wshanks replied:
> The reference key is important only when a user assigns a subroutine to reference.

The key also factors into the new parameter scoping features. If you call a schedule without passing a name, you will get difficult to use scoped parameter names.

One thing I wanted to check -- do you want to support calling different schedules with the same name passed explicitly? It sounds like the reference manager will raise an exception for the second schedule now whereas before the PR it was allowed to use the same name since the name was just attached to the instruction and not inserted into a shared namespace.

> keys = name or uuid.uuid4().hex

The problem with this is that the first key gets used for `Instruction.name` so it is probably best to keep that a readable string.

nkanazawa1989 replied:
Yes, this changes the behavior. I updated API doc and release note accordingly in https://github.com/Qiskit/qiskit-terra/pull/8005/commits/ffd2130f01b0ad9e0bcdc0a88d716b8104f01b2f

> The problem with this is that the first key gets used for Instruction.name so it is probably best to keep that a readable string.

These comments all make sense. I leave the current implementation.

nkanazawa1989 commented on qiskit/pulse/builder.py:
```diff
@@ -769,39 +784,46 @@ def call_subroutine(
+            for param in params:
+                local_assignment[param] = value
+        if value_dict:
+            local_assignment.update(value_dict)
```
Yes, there could be collision but this happens only intentionally. I added warning just in case. https://github.com/Qiskit/qiskit-terra/pull/8005/commits/dcf59a99647f05b902edd8851ccf371e162592bf

nkanazawa1989 commented:
Thanks @jakelishman you suggestions and comments were really useful to improve the code more reasonably and fix some bugs. I think your main concern is the hidden works at property methods, and this has been somewhat alleviated by turning `.scoped_parameters` into a method https://github.com/Qiskit/qiskit-terra/pull/8005/commits/5b7eeef21deec5d01ad0fd542275f135d54d7c55. The `.parameters` property still does the recursive call inside, but it's hard to change this without breaking API change.

jakelishman commented:
I'm not too bothered about the extra work done in properties, if you think it's appropriate here.  I personally don't like it (for the performance-concern reasons I cited above), but it's a valid and deliberate part of Python, and if you want to use it in your sections of the code that I don't maintain, that's totally fine and I wasn't at all planning to block review on it.

Using `property` to be able to avoid API breaks is one of the clearest good things about it, really.

Looks like there's a couple of test failures - I can re-review / approve once that's squashed.

nkanazawa1989 commented:
I plan to do major refactoring of the pulse module and we can convert these heavy-compute properties into methods with it. I think all of your comments have been addressed and the PR is good to go.

jakelishman reviewed (approved):
Thanks for the changes!  As long as all the pulse-specific people are happy with the interface, I'm happy for us to move this to merge.
